### PR TITLE
Introduce release file for java platform

### PR DESCRIPTION
**Pull request by fzakaria**
### Motivation
After haven spoken to @DarkDimius over slack regarding #2213
we discussed the option of creating a java platform release that
contains both linux & mac static binaries --
srb would be augmented to select the appropriate binary given the
`uname` of the system if the platform is java.

This a submission for a ruby script that will pull down pre-released
sorbet-static gems for linux & darwin and create a java one.

I don't have much insight into the BuildKite usage but my solution
proposed is one where the script is run following `publish-ruby-gems.sh`
for both platforms.

### Test plan
```
release_version=0.4.5083 .buildkite/build-static-release-java.rb
Downloaded sorbet-static-0.4.5083-universal-darwin-18
Unpacked gem: '/Users/faridzakaria/Development/sorbet/sorbet-static-0.4.5083-universal-darwin 18'
Downloaded sorbet-static-0.4.5083-x86_64-linux
Unpacked gem: '/Users/faridzakaria/Development/sorbet/sorbet-static-0.4.5083-x86_64-linux'
  Successfully built RubyGem
  Name: sorbet-static
  Version: 0.4.5083
  File: sorbet-static-0.4.5083-java.gem

tar --to-stdout -xf _out_/gems/sorbet-static-0.4.5083-java.gem data.tar.gz | tar -zt
libexec/linux.sorbet
libexec/mac.sorbet
```

### CODE REVIEW

**Comment by fzakaria** [comment]
I thought also about changing the filename to `build-and-publish-static-release-java.rb` and including the `gem push` statement at the end -- in the vein of it being a 1 stop-shop for publishing the java platform gem.

**Comment by jez** [comment]
> I thought also about changing the filename to `build-and-publish-static-release-java.rb` and including the `gem push` statement at the end -- in the vein of it being a 1 stop-shop for publishing the java platform gem.

It should definitely not both build and publish. We should never publish anything until we're sure that all the builds passed. For example, if it builds on linux but doesn't build on mac we shouldn't publish anything, because it means something platform specific is broken, and the end user is going to have no end of trouble trying to figure that out why something works on linux but is broken on mac when we already know the reason why from CI.

File: .buildkite/build-static-release-java.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+Dir.chdir("gems/sorbet-static") do
+  # build the java gem
+  # we cd otherwise gem can't find the files in the gemspec
+  system("gem build sorbet-static.gemspec", exception: true)
+
+end
+
+FileUtils.mkdir_p '_out_/gems'
+FileUtils.move "gems/sorbet-static/sorbet-static-#{release_version}-java.gem", '_out_/gems'
+
```
Comment (jez) [inline .buildkite/build-static-release-java.rb]: Why write this in Ruby? Every other build script is written in Bash.

Comment (jez) [inline .buildkite/build-static-release-java.rb]: Just adding this file won't have any effect on the build. If you want this file to run, you'll need to edit the `.buildkite/pipeline.yaml` file.

When you've edited that, you'll see the changes show up on the Buildkite build for your branch.

Comment (fzakaria) [reply .buildkite/build-static-release-java.rb]: oh -- i didn't know buildkite was public (https://buildkite.com/sorbet/sorbet)
I'll edit the pipeline.

Comment (fzakaria) [reply .buildkite/build-static-release-java.rb]: I'm not familiar enough to write something I would consider fault-tolerant in bash.
Seems simple enough to have written it in Ruby -- I hope that's ok.

**Comment by fzakaria** [comment]
@jez i don't have experience with BuildKite but I think the changes make sense.
I also changed the gem publishing step so it checks for each platform specifically -- the `gem list` methodology makes it too difficult to figure out if a particular platform a particular version is published.

ex. it puts the highest version for each platform on the single line
```
gem list --remote rubygems.org --exact 'sorbet-static'
sorbet-static (0.4.5095 universal-darwin-14 universal-darwin-15 universal-darwin-16 universal-darwin-17 universal-darwin-18 universal-darwin-19, 0.4.5083 x86_64-linux)
```

File: .buildkite/build-static-release-java.rb (hunk 1 above)
Comment (DarkDimius) [reply .buildkite/build-static-release-java.rb]: We're trying to be consistent about our environment and tools that we use for different steps of build.
Please redo this in Bash

Comment (DarkDimius) [reply .buildkite/build-static-release-java.rb]: > I'm not familiar enough to write something I would consider fault-tolerant in bash.

we'll help make it fault tolerant enough as part of review ;-) be brave, don't worry, we're here to help!

File: .buildkite/publish-ruby-gems.sh [hunk 2]
```diff
@@ -51,15 +51,28 @@ with_backoff() {
 # sorbet-static, but the sorbet-static gem push failed.
 #
 # (By failure here, we mean that RubyGems.org 502'd for some reason.)
-for gem_archive in "_out_/gems/sorbet-static-$release_version"-*.gem; do
-  if ! gem list --remote rubygems.org --exact 'sorbet-static' | grep -q "$release_version"; then
-    with_backoff gem push --verbose "$gem_archive"
+# Push the linux gem
+if ! gem fetch sorbet-static --platform x86_64-linux --version "$release_version" | grep -q "ERROR"; then
```
Comment (DarkDimius) [inline .buildkite/publish-ruby-gems.sh]: nice... this is a better fix than what we did in https://github.com/sorbet/sorbet/pull/2283

Comment (fzakaria) [reply .buildkite/publish-ruby-gems.sh]: thanks!

**Comment by fzakaria** [comment]
@DarkDimius @Jez it's in bash now

(:trollface: you'd think Sorbet would have all it's files in Ruby!)

File: .buildkite/build-static-release-java.rb (hunk 1 above)
Comment (fzakaria) [reply .buildkite/build-static-release-java.rb]: I appreciate the vote of confidence.
I meant it also slightly jokingly in that -- can any bash script be fault tolerant ;)

File: .buildkite/pipeline.yaml [hunk 3]
```diff
@@ -69,6 +69,11 @@ steps:
     artifact_paths: _out_/**/*
     <<: *elastic
 
+  - label: ":linux: build-java-ruby-gems.sh"
```
Comment (ptarjan) [inline .buildkite/pipeline.yaml]:
```suggestion
  - label: ":linux: build-static-release-java.sh"
```

**Comment by fzakaria** [comment]
@ptarjan can I get invited to buildkite organization to test running the new java platform build ? 
(farid.m.zakaria@gmail.com)

**Comment by ptarjan** [comment]
@fzakaria are you not able to view the results of your builds? they should be publicly visible

**Comment by fzakaria** [comment]
@ptarjan sorry -- i meant to kick off the pipeline once this gets merged in.
I can definitely see the pull-requests build results; I just upgraded the PR with some shellcheck fixes.

**Comment by ptarjan** [comment]
oh ok, well that should just auto-kick off for master when it is merged in.

Sadly we don't add folks from outside the core team to the buildkite right now. Do let us know if you need a build kicked off and we can help out

**Comment by fzakaria** [comment]
Okay -- @DarkDimius mentioned the publish step might only be once a day.
So I was hoping to just catch any potential mishaps quicker after hopefully this is merged.

I need to update `srb` afterwards to then handle the new `java` gem once I know it's publishing.

**Comment by ptarjan** [comment]
yeah, the actual gem publish is once a day. I can run a special one when i merge this

**Comment by fzakaria** [comment]
@ptarjan looks like the build failed because it doesn't run `build-static-release` on an OSX machine -- so the universal-darwin gem is missing.

**Comment by ptarjan** [comment]
I think we only run that step on master commits not on branches. Maybe your step should only run on master too if you want the output of that other step?

**Comment by fzakaria** [comment]
ah okay -- i didn't understand that configuration.
Let me update it to put the master in the master in the pipeline.

File: .buildkite/pipeline.yaml [hunk 4]
```diff
@@ -69,6 +69,14 @@ steps:
     artifact_paths: _out_/**/*
     <<: *elastic
 
+  - label: ":linux: build-static-release-java.sh"
```
Comment (ptarjan) [inline .buildkite/pipeline.yaml]:
```suggestion
  - label: ":linux: build-static-release-java.sh (master only)"
```

**Pull request merged**